ecommerce/extensions/payment/views/liqpay.py

```diff
@@ -0,0 +1,169 @@
+                transaction_id,
+                basket_id
+            )
+
+            basket = self._get_basket(basket_id)
+
+            if not basket:
+                logger.error('Received payment for non-existent basket [%s].', basket_id)
+                raise InvalidBasketError
+        finally:
+            # Store the response in the database.
+            ppr = self.payment_processor.record_processor_response(decode_data, transaction_id=transaction_id,
+                                                                   basket=basket)
+            logger.info("Successfully executed LiqPay payment [%s] for basket [%d].", transaction_id, basket.id)
+
+        receipt_url = get_receipt_page_url(
+            order_number=basket.order_number,
+            site_configuration=basket.site.siteconfiguration
+        )
+
+        try:
+            liqpay_response = request.POST.dict()
+            # Explicitly delimit operations which will be rolled back if an exception occurs.
+            with transaction.atomic():
+                try:
+                    self.handle_payment(liqpay_response, basket)
+                except InvalidSignatureError:
+                    logger.exception(
+                        'Received an invalid LiqPay response. The payment response was recorded in entry [%d].',
+                        ppr.id
+                    )
+                    return redirect(self.payment_processor.error_url)
+                except DuplicateReferenceNumber:
+                    if Order.objects.filter(number=order_number).exists() or PaymentProcessorResponse.objects.filter(
+                            basket=basket).exclude(transaction_id__isnull=True).exclude(transaction_id='').exists():
+                        logger.info(
+                            'Received LiqPay payment notification for basket [%d] which is associated '
+                            'with existing order [%s] or had an existing valid payment notification. '
+                            'No payment was collected, and no new order will be created.',
+                            basket.id,
+                            order_number
+                        )
+                    else:
+                        logger.info(
+                            'Received duplicate LiqPay payment notification for basket [%d] which is not associated '
+                            'with any existing order (Missing Order Issue)',
+                            basket.id,
+                        )
+                    return redirect(self.payment_processor.error_url)
+                except PaymentError:
+                    logger.exception(
+                        'LiqPay payment failed for basket [%d]. The payment response was recorded in entry [%d].',
+                        basket.id,
+                        ppr.id
+                    )
+                    return redirect(self.payment_processor.error_url)
```

Comment:
Too heavy for me and ugly. But up to you to refactor it.

Reply:
refactored